Feature/좋아요 기능 구현

rudeh1253 opened the pull request:
# 이슈번호
#42 

## 개요
- 좋아요 추가 및 취소 기능 구현
- 좋아요 관련 조회 기능 구현: 단일 조회, 특정 회원이 좋아요를 누른 목록 조회, 특정 글에 남겨진 좋아요 목록 조회, 특정 글에 달린 좋아요 개수 조회

## 추가사항
Post 엔티티에 Like에 대한 매핑을 추가하는 건 merge conflict 방지를 위해 내가 하지 않는다.


서비스하고 리포지토리는 짧은데 테스트 코드가 길어서 천 줄이 넘어가 버렸네요 ㅠ 죄송합니다...
코드양을 좀 더 신경써야겠네요...

rudeh1253 committed:
feat: 좋아요 관련 인터페이스 정의

Related to: #42

rudeh1253 committed:
test: Like 관련 비즈니스 로직 테스트 코드 작성

Related to: #42

rudeh1253 committed:
feat: 좋아요 관련 서비스 로직 구현

Related to: #42

rudeh1253 committed:
refactor(like): LikeService.addLike 메소드 수정

기존 코드에서는 isCanceled() == true인 엔티티에 대해서 addLike를 다시
호출할 경우 DuplicateLikeException이 발생. 이를 해결함

Related to: #42

rudeh1253 committed:
feat: 특정 Post에 남겨진 좋아요 개수 세는 메소드 구현

rudeh1253 committed:
test: 좋아요 API 테스트 코드 작성

Related to: #42

rudeh1253 committed:
feat: 좋아요 관련 Controller 및 ControllerAdvice 구현

Related to: #42

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/like/controller/LikeController.java:
```diff
@@ -0,0 +1,161 @@
+                    )
+            }
+    )
+    public ResponseEntity<ApiBasicResponse<LikeDto>> getSingleLike(@PathVariable("likeId") Long likeId) {
```
이 API가 필요한 이유가 무엇인가요?

rudeh1253 replied:
ID를 가지고 좋아요 단일 조회가 필요하다고 단순하게 생각해서 구현했는데 리뷰 보고 다시 생각해 보니 꼭 필요하진 않은 것 같습니다. 제가 안일하게 개발했습니다...
이 API는 삭제하고 나중에 필요하게 되면 다시 구현하겠습니다.

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/like/dto/LikeSearchDto.java:
```diff
@@ -0,0 +1,5 @@
+package com.hf.healthfriend.domain.like.dto;
+
+public class LikeSearchDto {
```
인자가 없는데 어디에 쓰이는 DTO인가요?

rudeh1253 replied:
앗 이거 삭제하려고 했는데 삭제 안 했네요
삭제하겠습니다!

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/like/entity/Like.java:
```diff
@@ -0,0 +1,63 @@
+    @JoinColumn(name = "post_id", nullable = false)
+    private Post post;
+
+    @Setter
```
Setter가 이 필드에서 사용되어야 하는 이유가 무엇인가요? cancel() / uncancel() 메소드에서 이미 this 키워드로 값은 변경하고 있지 않나요?

rudeh1253 replied:
맞습니다. 이 부분은 단순 실수입니다.
캐치해 주셔서 감사합니다! 삭제하겠습니다.

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/like/service/LikeService.java:
```diff
@@ -0,0 +1,89 @@
+                    new Member(memberId),
+                    Post.builder()
+                            .postId(postId)
+                            .build()
```
Like는 Member, Post 엔티티와 연관관계를 맺고 있으므로 이렇게 프록시 객체를 필드로 생성하는 것은 올바른 연관관계 매핑이 아니라 생각합니다. 단지 아이디를 담은 프록시 객체를 필드로 저장해놓은 것에 불과하다 생각합니다. 또 다른 문제점은 postId에 해당하는 post가 없는데 있는 것처럼 매핑될 수 있습니다. 개인적인 의견이니 반박 환영합니다!

rudeh1253 replied:
이 부분은 쿼리 호출을 최소화하기 위해 ID만 가지고 있는 엔티티 인스턴스를 생성했습니다. PostRepository와 MemberRepository에서 프록시 객체를 가지고 와서 Like 인스턴스를 생성할 때 생성자 파라미터로 넘겨준다면, Post와 Member 프록시 객체를 가져올 때 각각 쿼리를 추가적으로 호출해야 합니다. 데이터베이스 성능을 좀 더 신경쓰기 위해서 데이터베이스 접근 없이 ID만 가지고 있는 엔티티 인스턴스를 생성했습니다.

이 부분은 테스트를 돌리면서 로그를 확인함으로써 검증했습니다.

### ID만 가지고 있는 엔티티 인스턴스를 생성해서 Like 객체를 생성할 때 세팅할 경우
서비스 코드가 다음과 같이 작성되어 있을 때,
```java
// Autowired
private final LikeRepository likeRepository;

public Long addLike(Long memberId, Long postId) throws DuplicateLikeException {
    Optional<Like> likeOp = this.likeRepository.findByMemberIdAndPostId(memberId, postId);
    if (likeOp.isEmpty()) {
        Like like = new Like(
                new Member(memberId),
                Post.builder()
                        .postId(postId)
                        .build()
        );
        Like savedLike = this.likeRepository.save(like);
        return savedLike.getLikeId();
    }
    ...
}
```

다음과 같은 로그를 확인할 수 있습니다.
```
2024-09-26T18:42:28.066+09:00 DEBUG 27628 --- [           main] org.hibernate.SQL                        : 
    select
        l1_0.like_id,
        l1_0.is_canceled,
        l1_0.member_id,
        l1_0.post_id 
    from
        likes l1_0 
    where
        l1_0.post_id=? 
        and l1_0.member_id=?
Hibernate: 
    select
        l1_0.like_id,
        l1_0.is_canceled,
        l1_0.member_id,
        l1_0.post_id 
    from
        likes l1_0 
    where
        l1_0.post_id=? 
        and l1_0.member_id=?
2024-09-26T18:42:28.068+09:00 TRACE 27628 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BIGINT) <- [1]
2024-09-26T18:42:28.068+09:00 TRACE 27628 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (2:BIGINT) <- [1]
2024-09-26T18:42:28.076+09:00 DEBUG 27628 --- [           main] org.hibernate.SQL                        : 
    insert 
    into
        likes
        (is_canceled, member_id, post_id) 
    values
        (?, ?, ?)
Hibernate: 
    insert 
    into
        likes
        (is_canceled, member_id, post_id) 
    values
        (?, ?, ?)
2024-09-26T18:42:28.077+09:00 TRACE 27628 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BOOLEAN) <- [false]
2024-09-26T18:42:28.077+09:00 TRACE 27628 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (2:BIGINT) <- [1]
2024-09-26T18:42:28.077+09:00 TRACE 27628 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (3:BIGINT) <- [1]
```

여기서 Like 엔티티를 SELECT하는 쿼리 (Like 엔티티가 이미 존재하는지 확인)와, Like 엔티티를 INSERT하는 쿼리 두 개의 쿼리가 호출된 것을 확인할 수 있습니다.

### PostRepository와 MemberRepository로부터 프록시 객체를 가져올 경우
서비스 코드가 다음과 같이 작성되어 있을 때
```java
// Autowired
private final LikeRepository likeRepository;
private final PostRepository postRepository;
private final MemberRepository memberRepository;

public Long addLike(Long memberId, Long postId) throws DuplicateLikeException {
    Optional<Like> likeOp = this.likeRepository.findByMemberIdAndPostId(memberId, postId);
    if (likeOp.isEmpty()) {
        Member member = this.memberRepository.findById(memberId).get();
        Post post = this.postRepository.findById(postId).get();
        Like savedLike = this.likeRepository.save(new Like(member, post));
        return savedLike.getLikeId();
    }
    ...
}
```
다음과 같은 로그를 확인할 수 있습니다.
```
2024-09-26T18:41:08.520+09:00 DEBUG 33068 --- [           main] org.hibernate.SQL                        : 
    select
        l1_0.like_id,
        l1_0.is_canceled,
        l1_0.member_id,
        l1_0.post_id 
    from
        likes l1_0 
    where
        l1_0.post_id=? 
        and l1_0.member_id=?
Hibernate: 
    select
        l1_0.like_id,
        l1_0.is_canceled,
        l1_0.member_id,
        l1_0.post_id 
    from
        likes l1_0 
    where
        l1_0.post_id=? 
        and l1_0.member_id=?
2024-09-26T18:41:08.521+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BIGINT) <- [1]
2024-09-26T18:41:08.521+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (2:BIGINT) <- [1]
2024-09-26T18:41:08.542+09:00 DEBUG 33068 --- [           main] org.hibernate.SQL                        : 
    select
        m1_0.member_id,
        m1_0.birth_date,
        m1_0.companion_style,
        m1_0.creation_time,
        m1_0.email,
        m1_0.fitness_eagerness,
        m1_0.fitness_kind,
        m1_0.fitness_level,
        m1_0.fitness_objective,
        m1_0.gender,
        m1_0.introduction,
        m1_0.login_id,
        m1_0.nickname,
        m1_0.password,
        m1_0.profile_url,
        m1_0.role 
    from
        members m1_0 
    where
        m1_0.member_id=?
Hibernate: 
    select
        m1_0.member_id,
        m1_0.birth_date,
        m1_0.companion_style,
        m1_0.creation_time,
        m1_0.email,
        m1_0.fitness_eagerness,
        m1_0.fitness_kind,
        m1_0.fitness_level,
        m1_0.fitness_objective,
        m1_0.gender,
        m1_0.introduction,
        m1_0.login_id,
        m1_0.nickname,
        m1_0.password,
        m1_0.profile_url,
        m1_0.role 
    from
        members m1_0 
    where
        m1_0.member_id=?
2024-09-26T18:41:08.543+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BIGINT) <- [1]
2024-09-26T18:41:08.559+09:00 DEBUG 33068 --- [           main] org.hibernate.SQL                        : 
    select
        p1_0.post_id,
        p1_0.category,
        p1_0.content,
        p1_0.creation_time,
        p1_0.is_deleted,
        p1_0.last_modified,
        p1_0.writer_id,
        p1_0.title,
        p1_0.view_count 
    from
        post p1_0 
    where
        p1_0.post_id=?
Hibernate: 
    select
        p1_0.post_id,
        p1_0.category,
        p1_0.content,
        p1_0.creation_time,
        p1_0.is_deleted,
        p1_0.last_modified,
        p1_0.writer_id,
        p1_0.title,
        p1_0.view_count 
    from
        post p1_0 
    where
        p1_0.post_id=?
2024-09-26T18:41:08.559+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BIGINT) <- [1]
2024-09-26T18:41:08.565+09:00 DEBUG 33068 --- [           main] org.hibernate.SQL                        : 
    insert 
    into
        likes
        (is_canceled, member_id, post_id) 
    values
        (?, ?, ?)
Hibernate: 
    insert 
    into
        likes
        (is_canceled, member_id, post_id) 
    values
        (?, ?, ?)
2024-09-26T18:41:08.566+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (1:BOOLEAN) <- [false]
2024-09-26T18:41:08.566+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (2:BIGINT) <- [1]
2024-09-26T18:41:08.566+09:00 TRACE 33068 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter (3:BIGINT) <- [1]
```

여기서 like 테이블에 대한 SELECT 쿼리, post 테이블에 대한 SELECT 쿼리, member 테이블에 대한 SELECT 쿼리, like 테이블에 대한 INSERT 쿼리가 이루어지는 것을 확인할 수 있습니다.

저는 post와 member 테이블에 대한 SELECT 쿼리를 지양하기 위해 아이디만 가지고 있는 Post 엔티티와 Member 엔티티 인스턴스를 직접 생성해서 Like 엔티티에 세팅했습니다.

한편, 또 다른 문제점으로 제시해 주신 "postId에 해당하는 post가 없는데 있는 것처럼 매핑될 수 있"다는 점에 대해서 참조 무결성에 위배될 때 발생하는 예외를 처리하는 코드를 추가해야 할 것 같습니다. 이 점 지적 감사하며, 해당 예외를 처리하는 부분을 추가하겠습니다.

제 생각은 이러한데, 이 점에 대해서 가연님께서 또 어떻게 생각하시는지 궁금합니다! 좋은 지적 대단히 감사합니다. 덕분에 DB 접근과 연관관계 매핑에 대해 생각해 볼 수 있었습니다.

또한 제가 JPA에 대한 공부가 미진하여 올바른 연관관계 매핑가 아니라고 하신 부분에 대해 잘 와닿지가 않습니다.  간략하게 설명해주신다면 정말 감사하겠습니다! 제 실력이 부족하여 가연님께 수고를 감수하게 만들어 죄송합니다.

rudeh1253 replied:
찾아보니 JpaRepository에 가짜 프록시 객체를 만들어 주는 [getReferenceById](https://rachel0115.tistory.com/entry/Spring-Data-JPA-%EC%97%B0%EA%B4%80-%EA%B4%80%EA%B3%84%EB%A5%BC-%EA%B0%80%EC%A7%84-%EC%97%94%ED%8B%B0%ED%8B%B0%EB%A5%BC-%EC%A0%80%EC%9E%A5%ED%95%A0-%EB%95%8C-select-%EB%AC%B8%EC%9D%B4-%ED%95%84%EC%9A%94%ED%95%A0%EA%B9%8C-getOne-getReferenceById) 메소드가 있네요!

김영한 강사님 JPA 책을 읽을 때 본 적이 있는데 까맣게 잊고 있었습니다

단순히 id만 가지고 있는 가짜 프록시 객체를 직접 생성하는 것하고 getReferenceById를 사용하는 것 중에 뭐가 더 나은지 고민해봐야겠습니다

rudeh1253 reviewed (commented):
리뷰 확인했습니다! 리뷰해 주신 부분 반영한 뒤 다시 푸쉬해서 올리겠습니다!

rudeh1253 committed:
refactor(like): 코드 리뷰 반영하여 코드 수정

- likeId로 Like 단일 조회 엔드포인트 삭제
- Like 엔티티의 canceled에 붙어 있던 @Setter 삭제
- LikeSearchDto 삭제

Related to: #42
